File: server/csvParser.js
export function parseCSVBankStatement(csvText, filename = '') {
  try {
    if (!csvText || !csvText.trim()) {
      return { transactions: [], accountName: 'Imported account', extractedCount: 0 };
    }

    const lines = csvText.split(/\r?\n/).map(l => l.trim()).filter(Boolean);
    if (lines.length === 0) {
      return { transactions: [], accountName: 'Imported account', extractedCount: 0 };
    }

    // Helper to parse CSV line handling quotes
    const parseCSVLine = (lineStr) => {
      const result = [];
      let cur = '';
      let inQuotes = false;

      for (let i = 0; i < lineStr.length; i++) {
        const char = lineStr[i];
        if (char === '"') {
          inQuotes = !inQuotes;
        } else if (char === ',' && !inQuotes) {
          result.push(cur.trim());
          cur = '';
        } else {
          cur += char;
        }
      }
      result.push(cur.trim());
      return result;
    };

    // Find header row index (skips report title / metadata lines like Citi "Time period of report:")
    let headerRowIdx = 0;
    for (let i = 0; i < Math.min(10, lines.length); i++) {
      const parsedLine = parseCSVLine(lines[i]).map(h => h.toLowerCase().replace(/[^a-z0-9]/g, ''));
      const hasDate = parsedLine.some(h => h === 'date' || h.includes('date'));
      const hasDescOrAmount = parsedLine.some(h => h.includes('desc') || h.includes('amount') || h.includes('debit') || h.includes('merchant'));
      if (hasDate && hasDescOrAmount) {
        headerRowIdx = i;
        break;
      }
    }

    const header = parseCSVLine(lines[headerRowIdx]).map(h => h.toLowerCase().replace(/[^a-z0-9]/g, ''));

    // Find column indexes
    let dateIdx = header.findIndex(h => h.includes('transactiondate') || h.includes('postingdate') || h === 'date' || h.includes('date'));
    let descIdx = header.findIndex(h => h.includes('description') || h.includes('merchant') || h.includes('payee') || h.includes('name'));
    let categoryIdx = header.findIndex(h => h.includes('category'));
    let amountIdx = header.findIndex(h => h === 'amount' || h.includes('amount'));
    let debitIdx = header.findIndex(h => h === 'debit' || h.includes('debit'));
    let creditIdx = header.findIndex(h => h === 'credit' || h.includes('credit'));
    let typeIdx = header.findIndex(h => h === 'type' || h.includes('transactiontype') || h.includes('details'));
    let cardNoIdx = header.findIndex(h => h.includes('cardno') || h.includes('cardnum') || h.includes('account'));

    if (dateIdx === -1) dateIdx = 0;
    if (descIdx === -1) descIdx = 1;

    // Detect card number from data rows if cardNoIdx exists
    let extractedLast4 = '';
    if (cardNoIdx !== -1 && lines.length > headerRowIdx + 1) {
      const firstRow = parseCSVLine(lines[headerRowIdx + 1]);
      if (firstRow[cardNoIdx]) {
        const digits = firstRow[cardNoIdx].replace(/[^0-9]/g, '');
        if (digits.length >= 4) {
          extractedLast4 = digits.slice(-4);
        }
      }
    }

    // Determine account name from card number, filename, or header
    let accountName = 'Imported account';
    const lowerFilename = (filename || '').toLowerCase();
    const fullTextLower = csvText.slice(0, 300).toLowerCase();

    if (extractedLast4) {
      accountName = `Citi Card (...${extractedLast4})`;
    } else if (fullTextLower.includes('citi') || lowerFilename.includes('citi') || fullTextLower.includes('time period of report') || lowerFilename.includes('year to date') || lowerFilename.includes('annual account summary') || lowerFilename.includes('from 202')) {
      accountName = 'Citi Card';
    } else if (lowerFilename.includes('1554')) {
      accountName = 'Credit Card (...1554)';
    } else if (lowerFilename.includes('2067')) {
      accountName = 'Credit Card (...2067)';
    } else if (lowerFilename.includes('5589')) {
      accountName = 'Chase Private Client Checking (...5589)';
    } else if (lowerFilename.includes('3810')) {
      accountName = 'Bank of America Savings (...3810)';
    } else if (lowerFilename.includes('7387')) {
      accountName = 'Savings Account (...7387)';
    } else if (lowerFilename.includes('capital') || lowerFilename.includes('transaction_download')) {
      accountName = 'Capital One Card';
    } else if (lowerFilename.includes('checking')) {
      accountName = 'Checking Account';
    } else if (lowerFilename.includes('credit') || lowerFilename.includes('card')) {
      accountName = 'Credit Card';
    }

    const extractedTransactions = [];
    let endingBalance = null;

    const monthMap = { jan: '01', feb: '02', mar: '03', apr: '04', may: '05', jun: '06', jul: '07', aug: '08', sep: '09', oct: '10', nov: '11', dec: '12' };

    for (let i = headerRowIdx + 1; i < lines.length; i++) {
      const cols = parseCSVLine(lines[i]);
      if (cols.length <= Math.max(dateIdx, descIdx)) continue;

      const rawDate = cols[dateIdx];
      const rawDesc = cols[descIdx];
      const rawCategory = categoryIdx !== -1 ? cols[categoryIdx] : '';
      const rawType = typeIdx !== -1 ? cols[typeIdx] : '';

      if (!rawDate || !rawDesc) continue;

      // Extract amount and direction
      let numAmount = 0;
      let isIncome = false;

      if (debitIdx !== -1 && cols[debitIdx] && cols[debitIdx].trim()) {
        numAmount = parseFloat(cols[debitIdx].replace(/[^0-9.-]/g, ''));
        isIncome = false;
      } else if (creditIdx !== -1 && cols[creditIdx] && cols[creditIdx].trim()) {
        numAmount = parseFloat(cols[creditIdx].replace(/[^0-9.-]/g, ''));
        isIncome = numAmount > 0;
      } else if (amountIdx !== -1 && cols[amountIdx] && cols[amountIdx].trim()) {
        numAmount = parseFloat(cols[amountIdx].replace(/[^0-9.-]/g, ''));
        isIncome = numAmount > 0;
      }

      if (isNaN(numAmount) || numAmount === 0) continue;

      // Parse Date (MMM DD, YYYY or YYYY-MM-DD or MM/DD/YYYY)
      let isoDate = '2026-08-01';
      try {
        const rawDateTrim = rawDate.trim();
        const mmmMatch = rawDateTrim.match(/^([A-Za-z]{3})\s+(\d{1,2}),?\s+(\d{4})$/);
        
        if (mmmMatch) {
          const mKey = mmmMatch[1].toLowerCase();
          if (monthMap[mKey]) {
            const m = monthMap[mKey];
            const d = String(parseInt(mmmMatch[2], 10)).padStart(2, '0');
            const y = mmmMatch[3];
            isoDate = `${y}-${m}-${d}`;
          }
        } else if (/^\d{4}-\d{2}-\d{2}$/.test(rawDateTrim)) {
          isoDate = rawDateTrim;
        } else {
          const dParts = rawDateTrim.split(/[/.-]/);
          if (dParts.length === 3) {
            let y = parseInt(dParts[2] || dParts[0], 10);
            if (y < 100) y += 2000;
            const m = String(parseInt(dParts[0], 10)).padStart(2, '0');
            const d = String(parseInt(dParts[1], 10)).padStart(2, '0');
            if (y >= 2010 && y <= 2035) {
              isoDate = `${y}-${m}-${d}`;
            }
          }
        }
      } catch {}

      // Clean Merchant Name
      let merchant = rawDesc
        .replace(/[^\w\s*.-]/g, ' ')
        .replace(/\s+/g, ' ')
        .trim();

      if (!merchant || merchant.length < 2) continue;

      const lowerMerchant = merchant.toLowerCase();
      const lowerType = (rawType || '').toLowerCase();

      if (lowerType.includes('payment') || lowerMerchant.includes('citi autopay') || lowerMerchant.includes('payment thank you') || lowerMerchant.includes('mobile pymt')) {
        isIncome = true; // Credit Card Payment credit
      }

      const absAmount = Math.abs(numAmount);
      const type = isIncome ? 'income' : 'expense';

      // Auto-categorize
      let category = rawCategory && rawCategory.toLowerCase() !== 'other' ? rawCategory : (isIncome ? 'Income' : 'Needs review');
      if (lowerMerchant.includes('citi autopay') || lowerMerchant.includes('credit card payment') || lowerMerchant.includes('payment thank you')) {
        category = 'Credit Card Payment';
      } else if (lowerMerchant.includes('costco') || lowerMerchant.includes('walmart') || lowerMerchant.includes('kroger') || lowerMerchant.includes('grocers')) {
        category = 'Groceries';
      } else if (lowerMerchant.includes('lyft') || lowerMerchant.includes('uber')) {
        category = 'Transportation';
      } else if (lowerMerchant.includes('health') || lowerMerchant.includes('gohealth')) {
        category = 'Health & Fitness';
      } else if (lowerMerchant.includes('transfer') || lowerMerchant.includes('online transfer')) {
        category = 'Transfer';
      }

      extractedTransactions.push({
        date: isoDate,
        merchant,
        amount: absAmount,
        type,
        category,
        account: accountName,
        tags: ['CSV Import'],
        receipt: true,
        source: 'document'
      });
    }

    return {
      extractedCount: extractedTransactions.length,
      transactions: extractedTransactions,
      accountName,
      endingBalance
    };
  } catch (err) {
    console.error('CSV Parsing Error:', err.message);
    return {
      extractedCount: 0,
      transactions: [],
      accountName: 'Imported account',
      error: err.message
    };
  }
}

File: server/csv_parser.py
import csv
import io
import re
from datetime import datetime

MONTH_MAP = {
    'jan': '01', 'feb': '02', 'mar': '03', 'apr': '04', 'may': '05', 'jun': '06',
    'jul': '07', 'aug': '08', 'sep': '09', 'oct': '10', 'nov': '11', 'dec': '12'
}

def parse_csv_bank_statement(csv_text, filename=""):
    try:
        if not csv_text or not csv_text.strip():
            return {"transactions": [], "accountName": "Imported account", "extractedCount": 0}

        lines = [l.strip() for l in csv_text.splitlines() if l.strip()]
        if not lines:
            return {"transactions": [], "accountName": "Imported account", "extractedCount": 0}

        def parse_line(line_str):
            reader = csv.reader(io.StringIO(line_str))
            for row in reader:
                return [c.strip() for c in row]
            return []

        header_row_idx = 0
        for i in range(min(10, len(lines))):
            parsed = [re.sub(r'[^a-z0-9]', '', c.lower()) for c in parse_line(lines[i])]
            has_date = any('date' in c for c in parsed)
            has_desc_or_amt = any(('desc' in c or 'amount' in c or 'debit' in c or 'merchant' in c) for c in parsed)
            if has_date and has_desc_or_amt:
                header_row_idx = i
                break

        header = [re.sub(r'[^a-z0-9]', '', c.lower()) for c in parse_line(lines[header_row_idx])]

        def find_col(predicates):
            for p in predicates:
                for idx, h in enumerate(header):
                    if p(h):
                        return idx
            return -1

        date_idx = find_col([lambda h: 'transactiondate' in h, lambda h: 'postingdate' in h, lambda h: h == 'date', lambda h: 'date' in h])
        desc_idx = find_col([lambda h: 'description' in h, lambda h: 'merchant' in h, lambda h: 'payee' in h, lambda h: 'name' in h])
        category_idx = find_col([lambda h: 'category' in h])
        amount_idx = find_col([lambda h: h == 'amount', lambda h: 'amount' in h])
        debit_idx = find_col([lambda h: h == 'debit', lambda h: 'debit' in h])
        credit_idx = find_col([lambda h: h == 'credit', lambda h: 'credit' in h])
        type_idx = find_col([lambda h: h == 'type', lambda h: 'transactiontype' in h, lambda h: 'details' in h])
        card_no_idx = find_col([lambda h: 'cardno' in h, lambda h: 'cardnum' in h, lambda h: 'account' in h])

        if date_idx == -1: date_idx = 0
        if desc_idx == -1: desc_idx = 1

        extracted_last4 = ""
        if card_no_idx != -1 and len(lines) > header_row_idx + 1:
            first_row = parse_line(lines[header_row_idx + 1])
            if card_no_idx < len(first_row) and first_row[card_no_idx]:
                digits = re.sub(r'[^0-9]', '', first_row[card_no_idx])
                if len(digits) >= 4:
                    extracted_last4 = digits[-4:]

        account_name = "Imported account"
        lower_fn = filename.lower()
        full_text_lower = csv_text[:300].lower()

        if extracted_last4:
            account_name = f"Citi Card (...{extracted_last4})"
        elif 'citi' in full_text_lower or 'citi' in lower_fn or 'time period of report' in full_text_lower:
            account_name = "Citi Card"
        elif '1554' in lower_fn:
            account_name = "Credit Card (...1554)"
        elif '2067' in lower_fn:
            account_name = "Credit Card (...2067)"
        elif '5589' in lower_fn:
            account_name = "Chase Private Client Checking (...5589)"
        elif '3810' in lower_fn:
            account_name = "Bank of America Savings (...3810)"
        elif '7387' in lower_fn:
            account_name = "Savings Account (...7387)"
        elif 'capital' in lower_fn or 'transaction_download' in lower_fn:
            account_name = "Capital One Card"
        elif 'checking' in lower_fn:
            account_name = "Checking Account"
        elif 'credit' in lower_fn or 'card' in lower_fn:
            account_name = "Credit Card"

        extracted_transactions = []

        for i in range(header_row_idx + 1, len(lines)):
            cols = parse_line(lines[i])
            if len(cols) <= max(date_idx, desc_idx):
                continue

            raw_date = cols[date_idx] if date_idx < len(cols) else ""
            raw_desc = cols[desc_idx] if desc_idx < len(cols) else ""
            raw_cat = cols[category_idx] if (category_idx != -1 and category_idx < len(cols)) else ""
            raw_type = cols[type_idx] if (type_idx != -1 and type_idx < len(cols)) else ""

            if not raw_date or not raw_desc:
                continue

            num_amount = 0.0
            is_income = False

            if debit_idx != -1 and debit_idx < len(cols) and cols[debit_idx].strip():
                clean_num = re.sub(r'[^0-9.-]', '', cols[debit_idx])
                num_amount = float(clean_num) if clean_num else 0.0
                is_income = False
            elif credit_idx != -1 and credit_idx < len(cols) and cols[credit_idx].strip():
                clean_num = re.sub(r'[^0-9.-]', '', cols[credit_idx])
                num_amount = float(clean_num) if clean_num else 0.0
                is_income = num_amount > 0
            elif amount_idx != -1 and amount_idx < len(cols) and cols[amount_idx].strip():
                clean_num = re.sub(r'[^0-9.-]', '', cols[amount_idx])
                num_amount = float(clean_num) if clean_num else 0.0
                is_income = num_amount > 0

            if num_amount == 0.0:
                continue

            iso_date = "2026-08-01"
            try:
                raw_trim = raw_date.strip()
                mmm_match = re.match(r'^([A-Za-z]{3})\s+(\d{1,2}),?\s+(\d{4})$', raw_trim)
                if mmm_match:
                    m_key = mmm_match.group(1).lower()
                    if m_key in MONTH_MAP:
                        m_val = MONTH_MAP[m_key]
                        d_val = str(int(mmm_match.group(2))).zfill(2)
                        y_val = mmm_match.group(3)
                        iso_date = f"{y_val}-{m_val}-{d_val}"
                elif re.match(r'^\d{4}-\d{2}-\d{2}$', raw_trim):
                    iso_date = raw_trim
                else:
                    d_parts = re.split(r'[/.-]', raw_trim)
                    if len(d_parts) == 3:
                        y = int(d_parts[2] if len(d_parts[2]) == 4 else d_parts[0])
                        if y < 100: y += 2000
                        m = str(int(d_parts[0])).zfill(2)
                        d = str(int(d_parts[1])).zfill(2)
                        if 2010 <= y <= 2035:
                            iso_date = f"{y}-{m}-{d}"
            except Exception:
                pass

            merchant = re.sub(r'[^\w\s*.-]', ' ', raw_desc)
            merchant = re.sub(r'\s+', ' ', merchant).strip()
            if not merchant or len(merchant) < 2:
                continue

            lower_m = merchant.lower()
            lower_t = raw_type.lower()
            if 'payment' in lower_t or 'citi autopay' in lower_m or 'payment thank you' in lower_m or 'mobile pymt' in lower_m:
                is_income = True

            abs_amount = abs(num_amount)
            ttype = 'income' if is_income else 'expense'

            category = raw_cat if (raw_cat and raw_cat.lower() != 'other') else ('Income' if is_income else 'Needs review')
            if 'citi autopay' in lower_m or 'credit card payment' in lower_m or 'payment thank you' in lower_m:
                category = 'Credit Card Payment'
            elif any(k in lower_m for k in ['costco', 'walmart', 'kroger', 'grocers']):
                category = 'Groceries'
            elif 'lyft' in lower_m or 'uber' in lower_m:
                category = 'Transportation'
            elif 'health' in lower_m or 'gohealth' in lower_m:
                category = 'Health & Fitness'
            elif 'transfer' in lower_m or 'online transfer' in lower_m:
                category = 'Transfer'

            extracted_transactions.append({
                "date": iso_date,
                "merchant": merchant,
                "amount": abs_amount,
                "type": ttype,
                "category": category,
                "account": account_name,
                "tags": ['CSV Import'],
                "receipt": True,
                "source": 'document'
            })

        return {
            "extractedCount": len(extracted_transactions),
            "transactions": extracted_transactions,
            "accountName": account_name,
            "endingBalance": None
        }
    except Exception as e:
        return {
            "extractedCount": 0,
            "transactions": [],
            "accountName": "Imported account",
            "error": str(e)
        }
